### ISSUE TITLE
Consider not warning for `map_clone` with `Arc::clone`

### ISSUE THREAD

**Issue by stevenengler** [rust-lang/rust-clippy#12528]
### Summary

Similar to how [some people prefer](https://doc.rust-lang.org/std/sync/struct.Arc.html#deref-behavior) to use the fully qualified `Arc::clone(&foo)` rather than `foo.clone()` for readability, people may also prefer `foo.map(Arc::clone)` rather than `foo.cloned()` to be more explicit that the `Arc` is being cloned. I think that the `map_clone` lint should not trigger for `foo.map(Arc::clone)` or for `foo.map(Rc::clone)`.

This is of course subjective, but it would be a subjective choice to *not* show a warning, which I think is nicer.

### Lint Name

map_clone

### Reproducer

I tried this code:

```rust
use std::sync::Arc;

struct Foo;

fn main() {
    let x = Arc::new(Foo);
    let y = Some(&x);
    let _z = y.map(Arc::clone);
}
```

I saw this happen:

```text
warning: you are explicitly cloning with `.map()`
 --> src/main.rs:8:14
  |
8 |     let _z = y.map(Arc::clone);
  |              ^^^^^^^^^^^^^^^^^ help: consider calling the dedicated `cloned` method: `y.cloned()`
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#map_clone
  = note: `#[warn(clippy::map_clone)]` on by default
```

I expected to see this happen:

No lint.

### Version

```text
rustc 1.79.0-nightly (1388d7a06 2024-03-20)
binary: rustc
commit-hash: 1388d7a069d872bcfe5e5dd97ef61fa0a586fac0
commit-date: 2024-03-20
host: x86_64-unknown-linux-gnu
release: 1.79.0-nightly
LLVM version: 18.1.2
```

**Comment by y21** [comment]
That raises the question if we should also do this for `useless_asref then, which also currently lints on
```rs
opt /* : Option<Rc<_>> */
  .as_ref()
  .map(Rc::clone)
```
and suggests `.clone()` instead

**Comment by stevenengler** [comment]
Thanks for the quick change!

> That raises the question if we should also do this for useless_asref then, which also currently lints on

That's a good question, I didn't know about that lint.

My personal opinion is that it would be nicer to not lint for that example either, but it's a tougher argument since `opt.clone()` has significantly less clutter than `opt.as_ref().map(Rc::clone)`. I guess my feeling would be that default lints should have recommendations that are a clear win over the original code, and I'm not convinced that `opt.clone()` is a clear win over `opt.as_ref().map(Rc::clone)`. But I don't know what criteria the clippy team uses for lints.

If anyone wants me to open an issue for `useless_asref` to further the discussion, let me know.

**Comment by samueltardieu** [comment]
> That raises the question if we should also do this for `useless_asref` then, which also currently lints on
> 
> ```rust
> opt /* : Option<Rc<_>> */
>   .as_ref()
>   .map(Rc::clone)
> ```
> 
> and suggests `.clone()` instead

Yes, probably. I'll submit a PR to test this, after factoring the test.
